horovod/ray/strategy.py

```diff
@@ -175,4 +176,27 @@ def create_workers(self):
 
             self.workers.append(worker)
 
+        if self.use_gpu:
+            node_ids = ray.get(
+                [worker.node_id.remote() for worker in self.workers])
+            gpus = ray.get(
+                [worker.get_gpu_ids.remote() for worker in self.workers])
+            node_workers = defaultdict(list)
+            node_id_to_gpus = defaultdict(list)
+            for worker, node_id, worker_gpu_ids in zip(self.workers, node_ids,
+                                                       gpus):
+                node_workers[node_id].append(worker)
+                node_id_to_gpus[node_id].extend(worker_gpu_ids)
+
+            futures = []
+            for node_id, gpu_ids in node_id_to_gpus.items():
+                all_ids = ",".join([str(gpu_id) for gpu_id in gpu_ids])
+
+                for worker in node_workers[node_id]:
+                    futures.append(
+                        worker.update_env_vars.remote({
+                            "CUDA_VISIBLE_DEVICES":
+                            all_ids
```

Comment:
As of recent versions of NCCL, we no longer need to make sure that all visible devices on the node are visible to all local ranks, so it might make sense to remove this. Particularly because recent versions of TF and PyTorch will initialize memory on every GPU at startup, which can result in OOM errors if all devices are visible to all local ranks.

Reply:
ah yes, as mentioned offline: 

>  this is because Ray will set CUDA_VISIBLE_DEVICES to a single GPU on the node by default, and with Horovod we typically run with all visible GPUs

>  code that runs with all visible devices seems to be often used in examples and other estimator code